File: python/78.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        def dfs(nums):
            if len(nums)==1:
                res=[nums,[]]
                return res
            
            res=dfs(nums[1:])
            ans=[]
            for i in res:
                if i==None:
                    ans.append([nums[0]])
                    continue
                ans.append(i)
                
                ans.append(i+[nums[0]])
                
            return ans
        
        return dfs(nums)
class Solution(object):
    def subsets(self, nums):
        ret = []
        self.dfs(nums, [], ret)
        return ret
    
    def dfs(self, nums, path, ret):
        ret.append(path)
        for i in range(len(nums)):
            self.dfs(nums[i+1:], path+[nums[i]], ret)

File: java/binary_tree/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res= new ArrayList<>();
        if (nums==null) return res;
        res.add(new ArrayList<>());
        getResult(res,nums,0,new ArrayList<Integer>());
        return res;
        
    }
    private void getResult(List<List<Integer>> res,int[] nums, int level,ArrayList<Integer> lis){
        if (level==nums.length) return;
        for( int i=level;i<nums.length;i++){
            lis.add(nums[i]);
            res.add(new ArrayList<>(lis));
            getResult(res,nums,i+1,lis);
            lis.remove(lis.size()-1);
            
        }
        return;
        
    }
}